### PR TITLE
Fix #8920 - m4/1.4.19: Runtime assertion windows pops up on MSVC in Debug mode

### LINKED ISSUE #8920: [package] m4/1.4.19: Runtime assertion windows pops up on MSVC in Debug mode
### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **m4/1.4.19**
  * Operating System+version: **Windows 10**
  * Compiler+version: **MSVC 2017, 2019 in DEBUG**
  * Conan version: **conan 1.44.0**
  * Python version: **Python 3.9.5**

I have a recipe that depends on bison. 3.7.6 gives me parse errors, so I keep it at 3.7.1. I had pinned the exact bison/3.7.1 recipe and all worked well. I am trying to update my dependencies, so I tried the newest bison/3.7.1 revision, and I am now getting an assertion popup in m4.

I have tried various combinations to see if I could pinpoint what changed. My conclusion is that it's the move from m4/1.4.18 to m4/1.4.19 (I have tried the newest recipe revision of m4/1.4.18 and it worked, and the only m4/1.4.19 available fails).

```python
        # Latest bison/3.7.1 with m4/1.4.18
        #self.build_requires("bison/3.7.1#dcffa3dd9204cb79ac7ca09a7f19bb8b") #  Works

        # First bison/3.7.1 with m4/1.4.19
        self.build_requires("bison/3.7.1#47f49e709ddb9f8e055471c4e3c4e67d"): # Fails

        # Latest bison/3.7.1
        #self.build_requires("bison/3.7.1#ad29e804e82c8b6d58765096676b5a5e") # Fails
```

Screen Shot 2022-01-14 at 13 35 01

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)

```
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=Visual Studio
compiler.runtime=MDd
compiler.version=16
os=Windows
os_build=Windows
[options]
[build_requires]
[env]
```

### Steps to reproduce (Include if Applicable)

```
git clone git@github.com:NREL/conan-openstudio-ruby.git
git checkout update_ruby_installer_and_remotes

'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvarsall.bat' x64

conan create . openstudio_ruby/2.7.2@nrel/testing -b missing -b openstudio_ruby -s build_type=Debug
```

### Logs (Include/Attach if Applicable)

Example of a failed run: https://github.com/NREL/conan-openstudio-ruby/runs/4787415523?check_suite_focus=true

### PR DESCRIPTION
**Pull request by jmarrec**
Specify library name and version:  **m4/1.4.19**

Fix #8920 - m4/1.4.19: Runtime assertion windows pops up on MSVC in Debug mode by define HAVE_MSVC_INVALID_PARAMETER_HANDLER

---

-  [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by SSE4 — DISMISSED** [review]

File: recipes/m4/all/conanfile.py [hunk 1]
```diff
@@ -62,6 +62,9 @@ def _configure_autotools(self):
         elif self.settings.compiler == "clang":
             if tools.Version(self.version) < "1.4.19":
                 autotools.flags.extend(["-rtlib=compiler-rt", "-Wno-unused-command-line-argument"])
+        if self.settings.os == 'Windows':
```
Comment (SSE4) [inline recipes/m4/all/conanfile.py:65]: comparing compiler might be a better condition here, but that's fine as well

**Comment by madebr** [comment]
I just built with Visual Studio 2019 in Debug mode, but the assertion is still there.

**Comment by jmarrec** [comment]
I have no clue why it was working on my other machine the other day. Clearly this isn't working.

I've tweaked by patch-approach now, I was using the wrong macro. That seemed to work locally, so I pulled it here. Could you give it another test @madebr please?

I also wrote to the bug-m4 mailing list (hopefully clearly enough...)

**Comment by jmarrec** [comment]
My other machine has in profiles/default the compiler.runtime=MD  and not MDd, and that's why it was working I suppose.

**Comment by madebr** [comment]
It works now! Great job!

It's very weird though how a windows warning dialog sound is player every time after the test package has finished executing.

**Comment by jmarrec** [comment]
@madebr see the thread on the bug-m4 mailing list I have going: https://lists.gnu.org/archive/html/bug-m4/2022-01/msg00000.html

**Comment by madebr** [comment]
Thanks for the link. And +1 for the bug report, it looks like somebody is looking into it.

**Review by madebr — DISMISSED** [review]

**Comment by madebr** [comment]
@jmarrec 
Looks like they posted a patch: https://lists.gnu.org/archive/html/bug-m4/2022-01/txtSzXResQgKX.txt
It requires running gnulib.
To avoid that (since it takes a long of time for it to finish), the gnulib script can be run on your system and a patch created with the only the fclose diff (or at least reduced to the minimum).

**Comment by SSE4** [comment]
it would be nice if someone can test this patch:
```
 	xstrtoimax, xstrtoumax: depend on inttypes-incomplete
diff --git a/modules/close-stream b/modules/close-stream
index 78ed207992..4fce2ea233 100644
--- a/modules/close-stream
+++ b/modules/close-stream
@@ -6,6 +6,7 @@ lib/close-stream.h
 lib/close-stream.c
 
 Depends-on:
+fclose
 fpending
 stdbool
```
I'd prefer to have this one, so we diverge less from the upstream and don't need to constantly rebase custom patches

**Review by prince-chrismc — APPROVED** [review]

**Comment by jmarrec** [comment]
@SSE4 Based on the latest conversations on the mailing list, it seems the patch will have no effect. What they recommend is to (force) build with `/MD` instead of `/MDd`.

**Review by SSE4 — APPROVED** [review]

**Comment by SSE4** [comment]
`m4.1` (man page) depends on `m4.c` on the Makefile (which is modified by the patch).

**Review by madebr — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

File: recipes/m4/all/conanfile.py [hunk 2]
```diff
@@ -83,16 +83,19 @@ def _build_context(self):
...
 
     def _patch_sources(self):
         for patch in self.conan_data.get("patches", {}).get(self.version, []):
             tools.patch(**patch)
 
     def build(self):
+        with tools.chdir(self._source_subfolder):
+            tools.save("help2man", '#!/usr/bin/env bash\n:')
+            if os.name == 'posix':
+                os.chmod("help2man", os.stat("help2man").st_mode | 0o111)
```
Comment (jmarrec) [inline recipes/m4/all/conanfile.py:100]: Clever! thanks for the workaround @SSE4 !

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**